### PR TITLE
[CI] Provide latest FlatBuffers to the Engine build.

### LINKED ISSUE #5171: Update `Flatbuffers` toolchain
This task is automatically imported from the old Task Issue Board and it was originally created by Wojciech Daniło.
Original issue is [here](https://www.pivotaltracker.com/story/show/180303547).
<hr />

## Description
The Old Flatbufferes version has several problems and is hard to be installed on some systems.

## Resources
Flatbuffers has released [2.0.0](https://github.com/google/flatbuffers/releases/tag/v2.0.0) which overhauls rust support and fixes numerous bugs.

## Tasks:
- [ ] Upgrade both the engine and IDE to be compatible with the new `flatc` version.
- [ ] Split up the flatbuffers schemas into multiple organised files.

### PR DESCRIPTION
**Pull request by mwu-tow**
### Pull Request Description
This PR bumps the FlatBuffers version used by the backend to `24.3.25` (the latest version as of now).

Since the newer FlatBuffers releases come with prebuilt binaries for all platforms we target, we can simplify the build process by simply downloading the required `flatc` binary from the official FlatBuffers GitHub release page. This allows us to remove the dependency on `conda`, which was the only reliable way to get the outdated `flatc`.

The `conda` setup has been removed from the CI steps and the relevant code has been removed from the build script. 

The FlatBuffers version is no longer hard-coded in the Rust build script, it is inferred from the `build.sbt` definition (similar to GraalVM). 

### Important Notes

This does not affect the GUI binary protocol implementation.

While I initially wanted to update it, it turned out farly non-trivial.

As there are multiple issues with the generated TS code, it was significantly refactored by hand and it is impossible to automatically update it. Work to address this problem is left as [a future task](https://github.com/enso-org/enso/issues/9658).

As the Flatbuffers binary protocol is guaranteed to be compatible between versions (unlike the generated sources), there should be no adverse effects from bumping `flatc` only on the backend side.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: .github/workflows/std-libs-benchmark.yml [hunk 1]
```diff
@@ -18,12 +18,6 @@ jobs:
     runs-on:
       - benchmark
     steps:
-      - if: startsWith(runner.name, 'GitHub Actions') || startsWith(runner.name, 'Hosted Agent')
```
Comment (JaroslavTulach) [inline .github/workflows/std-libs-benchmark.yml:21]: Removing `conda` at so many places is great, but shouldn't there also be a removal of the usage of `code` somewhere?

File: build.sbt [hunk 2]
```diff
@@ -486,7 +486,7 @@ val zio = Seq(
 val bcpkixJdk15Version      = "1.70"
 val declineVersion          = "2.4.1"
 val directoryWatcherVersion = "0.18.0"
-val flatbuffersVersion      = "1.12.0"
+val flatbuffersVersion      = "24.3.25"
```
Comment (JaroslavTulach) [inline build.sbt:489]: Looks like a pretty recent version of `flatbuffers`. What will be the setup on my Ubuntu 22.04? Will I have to upgrade to new Ubuntu?

File: build/build/src/engine/context.rs [hunk 3]
```diff
@@ -146,35 +144,15 @@ impl RunContext {
...
-                }
-            }
+        // Setup flatc (FlatBuffers compiler), required for building the engine.
+        let flatc_goodie = cache::goodie::flatc::Flatc {
+            version:  engine::deduce_flatbuffers(&self.repo_root.build_sbt).await?,
+            platform: TARGET_OS,
+        };
+        flatc_goodie.install_if_missing(&self.cache).await?;
 
-            ide_ci::programs::Conda
```
Comment (JaroslavTulach) [inline build/build/src/engine/context.rs:169]: OK, so this is the code that was installing version `1.12.0` of `flatbuffers`.

File: build/ci_utils/src/cache/goodie/flatc.rs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+use crate::cache::goodie;
+use crate::cache::Cache;
+use crate::env::known::PATH;
+use crate::env::Modification;
+use crate::github::RepoRef;
+
+
+
+/// The repository of the FlatBuffers project.
+pub const REPO: RepoRef = RepoRef { owner: "google", name: "flatbuffers" };
```
Comment (JaroslavTulach) [inline build/ci_utils/src/cache/goodie/flatc.rs:12]: Now we are about to download `flatbuffers` from GitHub directly.

File: build/ci_utils/src/cache/goodie/flatc.rs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    /// # fn main() -> Result {
+    /// let version = Version::from_str("24.3.25")?;
+    /// let flatc = Flatc { version, platform: OS::Linux };
+    /// let url = flatc.url()?;
+    /// assert_eq!(url.as_str(), "https://github.com/google/flatbuffers/releases/download/v24.3.25/Linux.flatc.binary.clang++-15.zip");
+    /// # Ok(())
+    /// # }
+    pub fn url(&self) -> Result<Url> {
+        let filename = match (&self.version, self.platform) {
+            (version, OS::Windows) if version >= &Version::new(2, 0, 0) =>
```
Comment (JaroslavTulach) [inline build/ci_utils/src/cache/goodie/flatc.rs:40]: We do support multiple versions, but the only one used is `24.3.25`, right?

File: engine/language-server/src/main/schema/binary_protocol.fbs [hunk 6]
```diff
@@ -13,10 +13,10 @@ union InboundPayload {
 // An envelope for inbound requests and commands.
 table InboundMessage {
   // A unique id of the message sent to the server.
-  messageId: EnsoUUID (required);
+  message_id: EnsoUUID (required);
```
Comment (JaroslavTulach) [inline engine/language-server/src/main/schema/binary_protocol.fbs:16]: Looks like an incompatible, unnecessary change. Can we continue to use the old names of identifiers?

**Review by JaroslavTulach — CHANGES_REQUESTED** [review]
- Please continue to use camel case identifiers in `.fbs` file
- Update [CONTRIBUTING](https://github.com/enso-org/enso/blob/71427c3319edd8b46bc66b7cd645d837a66ab445/docs/CONTRIBUTING.md#system-requirements)
- can I get the right version of flatbuffers on Ubuntu without using `./run` easily?

File: engine/language-server/src/main/schema/binary_protocol.fbs (hunk 6 above)
Comment (mwu-tow) [reply engine/language-server/src/main/schema/binary_protocol.fbs:16]: The newer `flatc` version issue warnings when fields are not following `snake_case`, so we should change unless there is a good reason not to.

The compatibility is not affected. The binary protocol itself is agnostic of field names. In the generated code the `flatc` tries to follow whatever it considers as appropriate for given target language. In our case, the Java code still keeps the `camelCase`.

File: build/ci_utils/src/cache/goodie/flatc.rs (hunk 5 above)
Comment (mwu-tow) [reply build/ci_utils/src/cache/goodie/flatc.rs:40]: Yes. The Rust code for downloading the `flatc` binary supports all versions `>=2.0.0`, however the only one used is the one defined in `build.sbt`.
I used this previously for comparing the generated code between different `flatc` versions (when investigating bumping it for GUI as well) and thought there is no harm in keeping extra bit of leeway here.

File: build.sbt (hunk 2 above)
Comment (mwu-tow) [reply build.sbt:489]: For `./run`, it will automatically download the binary, so no additional setup is needed. 
If you avoid run, you'll have to provide the `flatc` binary on your own. The reliable way should be [downloading release asset](https://github.com/google/flatbuffers/releases/tag/v24.3.25) but if you have a package manager that happens to support that particular version, you can use it.

**Comment by mwu-tow** [comment]
> * Please continue to use camel case identifiers in `.fbs` file
> 
>     * Update [CONTRIBUTING](https://github.com/enso-org/enso/blob/71427c3319edd8b46bc66b7cd645d837a66ab445/docs/CONTRIBUTING.md#system-requirements)
> 
>     * can I get the right version of flatbuffers on Ubuntu without using `./run` easily?

I have updated the `CONTRIBUTING.md`. As to other points, I've replied in the conversation comments.

**Review by Akirathan — APPROVED** [review]

**Review by JaroslavTulach — APPROVED** [review]

File: build/build/src/lib.rs [hunk 7]
```diff
@@ -99,6 +99,11 @@ pub fn get_graal_packages_version(build_sbt_contents: &str) -> Result<Version> {
     get_string_assignment_value(build_sbt_contents, "graalMavenPackagesVersion")?.parse2()
 }
 
+/// Get version of GraalVM packages from the `build.sbt` file contents.
+pub fn get_flatbuffers_version(build_sbt_contents: &str) -> Result<Version> {
+    get_string_assignment_value(build_sbt_contents, "flatbuffersVersion")?.parse2()
```
Comment (hubertp) [inline build/build/src/lib.rs:104]: We should think of using some tool agnostic format (YAML?) for storing library versions

**Review by hubertp — APPROVED** [review]
Glad to see conda gone.

Should update `docs/CONTRIBUTING` not to install `flatc` via `conda`.

**Review by radeusgd — APPROVED** [review]
